pipe_file fails to invalidate directory cache

The sequence of calls `ls -> pipe_file -> ls` causes the newly written file to not be found in the second `ls`. Here's a failing test (using Zarr's test setup, which I think was taken from s3fs).

```python
async def test_invalidate_cache(s3: s3fs.S3FileSystem) -> None:
    await s3._call_s3("put_object", Bucket=test_bucket_name, Key="a/b.txt", Body=b"abc")
    before = await s3._ls(f"{test_bucket_name}/a/")
    assert sorted(before) == ["test/a/b.txt"]
    await s3._pipe_file(f"{test_bucket_name}/a/c.txt", data=b"abc")
    after = await s3._ls(f"{test_bucket_name}/a/")
    assert sorted(after) == ["test/a/b.txt", "test/a/c.txt"]
```

That fails with

```
    async def test_invalidate_cache(s3: s3fs.S3FileSystem) -> None:
        await s3._call_s3("put_object", Bucket=test_bucket_name, Key="a/b.txt", Body=b"abc")
        before = await s3._ls(f"{test_bucket_name}/a/")
        assert sorted(before) == ["test/a/b.txt"]
        await s3._pipe_file(f"{test_bucket_name}/a/c.txt", data=b"abc")
        after = await s3._ls(f"{test_bucket_name}/a/")
>       assert sorted(after) == ["test/a/b.txt", "test/a/c.txt"]
E       AssertionError: assert ['test/a/b.txt'] == ['test/a/b.tx...test/a/c.txt']
E         
E         Right contains one more item: 'test/a/c.txt'
E         
E         Full diff:
E           [
E               'test/a/b.txt',
E         -     'test/a/c.txt',
E           ]

tests/test_store/test_remote.py:275: AssertionError
```

I think the issue is in`pipe_file`, which has an early return at https://github.com/fsspec/s3fs/blob/dd75a1a076d176049ff1f3d9f616f1a724f794df/s3fs/core.py#L1141-L1143. The `self.invalidate_cache` at the end of the function isn't run.

xref https://github.com/pydata/xarray/issues/9515

--- a/s3fs/core.py
+++ b/s3fs/core.py
@@ -1141,9 +1141,11 @@ class S3FileSystem(AsyncFileSystem):
         size = len(data)
         # 5 GB is the limit for an S3 PUT
         if size < min(5 * 2**30, 2 * chunksize):
-            await self._call_s3(
+            out = await self._call_s3(
                 "put_object", Bucket=bucket, Key=key, Body=data, **kwargs
             )
+            self.invalidate_cache(path)
+            return out
         else:
 
             mpu = await self._call_s3(
@@ -1177,7 +1179,7 @@ class S3FileSystem(AsyncFileSystem):
                 UploadId=mpu["UploadId"],
                 MultipartUpload={"Parts": parts},
             )
-        self.invalidate_cache(path)
+            self.invalidate_cache(path)
 
     async def _put_file(
         self,

--- a/s3fs/core.py
+++ b/s3fs/core.py
@@ -1141,7 +1141,7 @@ class S3FileSystem(AsyncFileSystem):
         size = len(data)
         # 5 GB is the limit for an S3 PUT
         if size < min(5 * 2**30, 2 * chunksize):
-            return await self._call_s3(
+            await self._call_s3(
                 "put_object", Bucket=bucket, Key=key, Body=data, **kwargs
             )
         else:

--- a/s3fs/tests/test_s3fs.py
+++ b/s3fs/tests/test_s3fs.py
@@ -2802,3 +2802,13 @@ def test_upload_part_with_prime_pads(s3_fixed_upload_size):
 
     with s3_fixed_upload_size.open(a, "r") as f:
         assert len(f.read()) == 2 * block + pad1 + pad2
+
+
+@pytest.mark.asyncio
+async def test_invalidate_cache(s3: s3fs.S3FileSystem) -> None:
+    await s3._call_s3("put_object", Bucket=test_bucket_name, Key="a/b.txt", Body=b"abc")
+    before = await s3._ls(f"{test_bucket_name}/a/")
+    assert sorted(before) == ["test/a/b.txt"]
+    await s3._pipe_file(f"{test_bucket_name}/a/c.txt", data=b"abc")
+    after = await s3._ls(f"{test_bucket_name}/a/")
+    assert sorted(after) == ["test/a/b.txt", "test/a/c.txt"]

